File: LLD/synchronization_with_semaphores/ZeroEvenOdd/ZeroEvenOdd.java
package LLD.synchronization_with_semaphores.ZeroEvenOdd;

import java.util.concurrent.Semaphore;
import java.util.function.IntConsumer;

/*
 * You have an object printNumber. printNumber.accept(x) can be called with an integer parameter that prints it to the console.

You are given an instance of the class ZeroEvenOdd that has three functions: zero, even, and odd. The same instance of ZeroEvenOdd will be passed to three different threads:

Thread A: calls zero() that should only output 0's.
Thread B: calls even() that should only output even numbers.
Thread C: calls odd() that should only output odd numbers.
Modify the given class to output the series "010203040506..." where the length of the series must be 2n.

Implement the ZeroEvenOdd class:

ZeroEvenOdd(int n) Initializes the object with the number n that represents the numbers that should be printed.
void zero(printNumber) Calls printNumber to output one zero.
void even(printNumber) Calls printNumber to output one even number.
void odd(printNumber) Calls printNumber to output one odd number.
Example 1:
Input: n = 2
Output: "0102"
Explanation: There are three threads being fired asynchronously.
One of them calls zero(), the other calls even(), and the last one calls odd().
"0102" is the correct output.
Example 2:
Input: n = 5
Output: "0102030405"
Constraints:
1 <= n <= 1000
 */
class ZeroEvenOdd {
    private int n;
    private Semaphore zeroSemaphore;
    private Semaphore oddSemaphore;
    private Semaphore evenSemaphore;

    public ZeroEvenOdd(int n) {
        this.n = n;
        this.zeroSemaphore = new Semaphore(1);  // Start with zero
        this.oddSemaphore = new Semaphore(0);   // Odd waits
        this.evenSemaphore = new Semaphore(0);  // Even waits
    }

    public void zero(IntConsumer printNumber) throws InterruptedException {
        for (int i = 1; i <= n; i++) {
            zeroSemaphore.acquire();
            printNumber.accept(0); // Print 0
            if (i % 2 == 1) {
                oddSemaphore.release();  // Allow odd
            } else {
                evenSemaphore.release(); // Allow even
            }
        }
    }

    public void odd(IntConsumer printNumber) throws InterruptedException {
        for (int i = 1; i <= n; i += 2) {
            oddSemaphore.acquire();
            printNumber.accept(i); // Print odd number
            zeroSemaphore.release(); // Allow next 0
        }
    }

    public void even(IntConsumer printNumber) throws InterruptedException {
        for (int i = 2; i <= n; i += 2) {
            evenSemaphore.acquire();
            printNumber.accept(i); // Print even number
            zeroSemaphore.release(); // Allow next 0
        }
    }
}

File: LLD/synchronization_with_semaphores/ZeroEvenOdd/ZeroEvenOdd.py
# Zero Even Odd Challenge
# Problem Statement
# In this challenge, you are required to design a multithreading solution that ensures the ordered printing of numbers in a zero-even-odd sequence up to a given n. Specifically, you'll be working with a class ZeroEvenOdd that provides three methods:

# zero(printNumber): prints "0" to the standard output.
# even(printNumber): prints even numbers to the standard output.
# odd(printNumber): prints odd numbers to the standard output.
# Each of these methods is intended to be executed by a different thread.

# Task Description
# Your task is to implement the ZeroEvenOdd class to ensure that:

# The zero method prints "0" and is called n times.
# The even method prints even numbers from 2 to n.
# The odd method prints odd numbers from 1 to n.
# The numbers are printed in sequence, with "0" printed between each number, resulting in a pattern like "0102030405" for n = 5.
# Requirements
# Utilize semaphores from Python's threading module to synchronize the execution order of zero, even, and odd methods.
# The PrintNumber class is provided to encapsulate the logic for printing numbers and should be used as is.
# Ensure your implementation allows the threads to cooperate smoothly, maintaining the correct sequence regardless of how they are scheduled by the Python interpreter.
# Implementation Details
# The ZeroEvenOdd class constructor accepts an integer n, representing the maximum number to print.
# You need to correctly initialize and use semaphores to control the execution flow between the threads running zero, even, and odd methods.

import threading

class ZeroEvenOdd:
    def __init__(self, n):
        self.n = n
        self.zero_sem = threading.Semaphore(1)  # Zero starts first
        self.odd_sem = threading.Semaphore(0)   # Odd waits
        self.even_sem = threading.Semaphore(0)  # Even waits

    def zero(self, printNumber):
        for i in range(1, self.n + 1):
            self.zero_sem.acquire()  # Wait for zero
            printNumber.accept(0)  # Print "0"
            if i % 2 == 1:
                self.odd_sem.release()  # Allow odd to print
            else:
                self.even_sem.release()  # Allow even to print

    def even(self, printNumber):
        for i in range(2, self.n + 1, 2):
            self.even_sem.acquire()  # Wait for zero
            printNumber.accept(i)  # Print even number
            self.zero_sem.release()  # Allow zero to print again

    def odd(self, printNumber):
        for i in range(1, self.n + 1, 2):
            self.odd_sem.acquire()  # Wait for zero
            printNumber.accept(i)  # Print odd number
            self.zero_sem.release()  # Allow zero to print again


class PrintNumber:
    def accept(self, x):
        print(x, end='')


def test_zero_even_odd():
    n = 8
    zeo = ZeroEvenOdd(n)
    pn = PrintNumber()

    threads = [
        threading.Thread(target=zeo.zero, args=(pn,)),
        threading.Thread(target=zeo.even, args=(pn,)),
        threading.Thread(target=zeo.odd, args=(pn,))
    ]

    for t in threads:
        t.start()

    for t in threads:
        t.join()


if __name__ == "__main__":
    test_zero_even_odd()
